### PR TITLE
User login

### PR DESCRIPTION
**Pull request by mferenca**
Please review @marjev

### CODE REVIEW

File: acceptance_tests/mixins.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+from urlparse import urljoin
+from django.test import TestCase
```
Comment (marjev) [inline acceptance_tests/mixins.py]: Add an empty line between the first two imports.

File: acceptance_tests/mixins.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+from selenium import webdriver
```
Comment (marjev) [inline acceptance_tests/mixins.py]: django and selenium should be grouped together.

File: acceptance_tests/test_login_page.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+from mixins import BaseTestCase
+
+from page_objects import PageObject, PageElement
```
Comment (marjev) [inline acceptance_tests/test_login_page.py]:
```
from page_objects import PageObject, PageElement

from .mixins import BaseTestCase
```

File: acceptance_tests/test_register_page.py [hunk 4]
```diff
@@ -1,28 +1,24 @@
-from urlparse import urljoin
-from django.test import TestCase
+from mixins import BaseTestCase
 
 from page_objects import PageObject, PageElement
-from selenium import webdriver
```
Comment (marjev) [inline acceptance_tests/test_register_page.py]: Order of the imports is wrong. See the previous comment.

File: users/forms.py [hunk 5]
```diff
@@ -24,3 +24,14 @@ def save(self, commit=True):
             user.save()
 
         return user
+
+
+class LoginForm(AuthenticationForm):
+
```
Comment (marjev) [inline users/forms.py]: nitpick: you don't need this line.

File: users/static/users/login-register.css [hunk 6]
```diff
@@ -60,32 +84,59 @@ h1 {
...
+.login-register-button.secondary:hover {
+    transition: all 0.5s ease;
+    box-shadow: 0 2px 2px 0 rgba(0,0,0,0.14), 0 3px 1px -2px rgba(0,0,0,0.12), 0 1px 5px 0 rgba(0,0,0,0.2);
+    background-color: #4285f4;
+    color: #ffffff;
+}
+
 .note {
     margin: 12px 0 0;
     text-align: center;
```
Comment (marjev) [inline users/static/users/login-register.css:142]: Re-check the css properties to make sure they're consistent with css concentric style or if you want, you can find some other guidelines for writing CSS which feels more natural to you.

It's to avoid having css styles that are hard to maintain.

File: users/templates/login-register.html [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+<body>
+  <div class="login-register-container">
+    <div class="container-primary">
+      <h1>{% block page-title %}{% endblock %}</h1>
+      <form method="post">
+        {% csrf_token %}
+        {% if messages %}
+        <ul class="message-success">
+          {% for message in messages %}
+          <li>{{ message }}</li>
```
Comment (marjev) [inline users/templates/login-register.html]: nitpick: indent <li> since it's within a loop.

File: users/templates/login-register.html [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+              <p class="form-error">{{ err }}</p>
+            {% endfor %}
+          </div>
+        {% endif %}
+        {% for field in form %}
+          <div class="field-wrapper">
+            <ul class="errorlist" id="{{ field.name }}-error">
+            {% for error in field.errors %}
+              <li>{{ error }}</li>
+            {% endfor %}
```
Comment (marjev) [inline users/templates/login-register.html]: nitpick: add indentation to for and endfor and additional one for li.

File: users/templates/user-profile.html [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+<html lang="en">
+<head>
+  <meta charset="UTF-8">
+  <title>User profile</title>
+</head>
+<body>
+  <p>You are logged in!</p>
+  <a href="{% url 'users:logout' %}">Logout</a>
+</body>
+</html>
```
Comment (marjev) [inline users/templates/user-profile.html]: Can you add another abstraction for templates?

You currently have this template and registration+login base template which is a parent for login and registration templates.
How about making a base template that has common html? Try DRYing this as much as you can to avoid retyping the same code over and over.

File: users/tests/tests.py [hunk 10]
```diff
@@ -4,39 +4,86 @@
...
+    def test_register_form_is_invalid(self):
         form_data = {'username': 'testUser', 'password1': 'pass1234', 'password2': 'pass123'}
         form = RegistrationForm(data=form_data)
         self.assertFalse(form.is_valid())
 
 
+class LoginFormUnitTests(TestCase):
+    def test_login_form_is_valid(self):
+        UserFactory(username='existing_user')
+        form_data = {'username': 'existing_user', 'password': 'pass1234'}
```
Comment (marjev) [inline users/tests/tests.py]: No need for this variable since it's never reused. You can send the value directly to the form.

File: users/tests/tests.py [hunk 11]
```diff
@@ -4,39 +4,86 @@
...
+        form = LoginForm(data=form_data)
+        self.assertTrue(form.is_valid())
+
+    def test_login_form_is_invalid(self):
+        form_data = {'username': 'non_existing_user', 'password': 'pass1234'}
```
Comment (marjev) [inline users/tests/tests.py]: No need for this variable since it's never reused. You can send the value directly to the form.

File: users/tests/tests.py [hunk 12]
```diff
@@ -4,39 +4,86 @@
...
+
 class UserRegistrationIntegrationTests(TestCase):
     def test_redirection_on_register_successful(self):
         response = self.client.post(reverse('users:register'), {'username': 'user_1', 'password1': 'pass1234', 'password2': 'pass1234'})
         self.assertEqual(response.status_code, 302)
 
     def test_register_successful(self):
         response = self.client.post(reverse('users:register'), {'username': 'user_1', 'password1': 'pass1234', 'password2': 'pass1234'}, follow=True)
-        self.assertIn('You successfully created a new account!', response.content)
+        self.assertIn('You successfully created your account, now you can sing in!', response.content)
```
Comment (marjev) [inline users/tests/tests.py]: You should import `UserRegistration` and use `success_message`. In case it changes, you want have to change this test.

File: users/urls.py [hunk 13]
```diff
@@ -4,6 +4,8 @@
 
 
 urlpatterns = [
+    url(r'^login/$', views.UserLogin.as_view(), name='login'),
     url(r'^register/$', views.UserRegistration.as_view(), name='register'),
-    url(r'^register-successful', views.UserRegistrationSuccessful.as_view(), name='register-successful')
+    url(r'^user-profile', views.UserProfilePage.as_view(), name='user-profile'),
+    url(r'^logout/$', views.Logout.as_view(), name='logout')
```
Comment (marjev) [inline users/urls.py]: alphabetize to make it easier to maintain.

File: users/tests/tests.py [hunk 14]
```diff
@@ -4,39 +4,86 @@
...
         UserFactory(username='existing_user')
         response = self.client.post(reverse('users:register'), {'username': 'existing_user', 'password1': 'pass1234', 'password2': 'pass1234'})
         self.assertIn('A user with that username already exists.', response.content)
+
+
+class UserLoginIntegrationTests(TestCase):
+    def test_redirection_on_login_successful(self):
+        UserFactory(username='existing_user')
+        response = self.client.post(reverse('users:login'), {'username': 'existing_user', 'password': 'pass1234'})
+        self.assertEqual(response.status_code, 302)
```
Comment (marjev) [inline users/tests/tests.py]: Also check where the user is being redirected to.
```
self.assertTrue(response['Location'].endswith(reverse('url')))
```
Just replace url with a valid value.

File: users/tests/tests.py [hunk 15]
```diff
@@ -4,39 +4,86 @@
...
+
+    def test_login_successful(self):
+        UserFactory(username='existing_user')
+        response = self.client.post(reverse('users:login'), {'username': 'existing_user', 'password': 'pass1234'}, follow=True)
+        self.assertIn('You are logged in!', response.content)
```
Comment (marjev) [inline users/tests/tests.py]: Don't follow, you can check if the user is redirected and where to.

File: users/tests/tests.py [hunk 16]
```diff
@@ -4,39 +4,86 @@
...
+
+    def test_login_unsuccessful(self):
+        response = self.client.post(reverse('users:login'), {'username': 'non_existing_user', 'password': 'pass1234'})
+        self.assertIn('Please enter a correct username and password. Note that both fields may be case-sensitive.', response.content)
```
Comment (marjev) [inline users/tests/tests.py]: Same comment as above.

File: users/tests/tests.py [hunk 17]
```diff
@@ -4,39 +4,86 @@
...
+
+
+class UserProfileIntegrationTests(TestCase):
+    def test_user_profile_page_with_authenticated_user(self):
+        UserFactory(username='existing_user')
+        self.client.post(reverse('users:login'), {'username': 'existing_user', 'password': 'pass1234'})
+        response = self.client.get(reverse('users:user-profile'))
+        self.assertIn('You are logged in!', response.content)
```
Comment (marjev) [inline users/tests/tests.py]: Ditto.

File: users/tests/tests.py [hunk 18]
```diff
@@ -4,39 +4,86 @@
...
+
+
+class AnonymousUserIntegrationTests(TestCase):
+    def test_redirection_on_anonymous_user_access_profile_page(self):
+        response = self.client.get(reverse('users:user-profile'))
+        self.assertEqual(response.status_code, 302)
```
Comment (marjev) [inline users/tests/tests.py]: Ditto.

File: users/tests/tests.py [hunk 19]
```diff
@@ -4,39 +4,86 @@
...
+
+    def test_redirection_to_login_on_anonymous_user_access_profile_page(self):
+        response = self.client.get(reverse('users:user-profile'), follow=True)
+        self.assertIn('Sign in', response.content)
```
Comment (marjev) [inline users/tests/tests.py]: Ditto.

File: users/views.py [hunk 20]
```diff
@@ -1,14 +1,39 @@
...
+    def get_success_url(self):
+        return reverse('users:login')
 
-class UserRegistrationSuccessful(TemplateView):
-    template_name = 'registration_successful.html'
+
+class UserLogin(FormView):
+    form_class = LoginForm
+    template_name = 'login.html'
+    success_url = reverse_lazy('users:user-profile')
```
Comment (marjev) [inline users/views.py]: nitpick: alphabetize

File: users/views.py [hunk 21]
```diff
@@ -1,14 +1,39 @@
...
-from users.forms import RegistrationForm
+from users.forms import LoginForm, RegistrationForm
 
 
-class UserRegistration(CreateView):
+class UserRegistration(SuccessMessageMixin, CreateView):
     form_class = RegistrationForm
     template_name = 'registration.html'
-    success_url = reverse_lazy('users:register-successful')
+    success_message = 'You successfully created your account, now you can sing in!'
```
Comment (marjev) [inline users/views.py]: nitpick: alphabetize

**Review by marjev — CHANGES_REQUESTED** [review]

**Comment by mferenca** [comment]
Addressed comments @marjev

File: acceptance_tests/mixins.py [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+
+
+class BaseTestCase(TestCase):
+    HOST = 'http://127.0.0.1:8080/'
+
+    def setUp(self):
+        self.driver = webdriver.Chrome()
+        self.login_url = urljoin(self.HOST, 'users/login')
+        self.register_url = urljoin(self.HOST, 'users/register')
+        self.user_profile_url = urljoin(self.HOST, 'users/user-profile')
```
Comment (marjev) [inline acceptance_tests/mixins.py]: Not all test cases care for these URLs. Define each of them in the class that uses them at the class level, like the HOST is defined within this class.

Comment (marjev) [reply acceptance_tests/mixins.py]: Also, can you use named URLs instead of hard coding them?

File: users/templates/login-register.html [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+    <div class="container-primary">
+      <h1>{% block page-title %}{% endblock %}</h1>
+      <form method="post">
+        {% csrf_token %}
+        {% if messages %}
+        <ul class="message-success">
+          {% for message in messages %}
+            <li>{{ message }}</li>
+          {% endfor %}
+        </ul>
```
Comment (marjev) [inline users/templates/login-register.html]: nitpick: `ul` should be indented since it's a child of `if` tag.

File: users/tests/tests.py [hunk 24]
```diff
@@ -34,9 +44,43 @@ def test_redirection_on_register_successful(self):
...
-        self.assertIn('You successfully created a new account!', response.content)
+        self.assertIn(UserRegistration.success_message, response.content)
 
     def test_register_unsuccessful(self):
         UserFactory(username='existing_user')
         response = self.client.post(reverse('users:register'), {'username': 'existing_user', 'password1': 'pass1234', 'password2': 'pass1234'})
         self.assertIn('A user with that username already exists.', response.content)
+
+
+class UserLoginIntegrationTests(TestCase, SimpleTestCase):
```
Comment (marjev) [inline users/tests/tests.py]: `SimpleTestCase` is a sub-class of the `TestCase`. You should be able to remove the `TestCase` without it having any impact on the tests.

File: users/tests/tests.py [hunk 25]
```diff
@@ -34,9 +44,43 @@ def test_redirection_on_register_successful(self):
...
+
+
+class AnonymousUserIntegrationTests(TestCase, SimpleTestCase):
+    def test_redirection_on_anonymous_user_access_profile_page(self):
+        response = self.client.get(reverse('users:user-profile'))
+        self.assertEqual(response.status_code, 302)
+
+    def test_redirection_to_login_on_anonymous_user_access_profile_page(self):
+        response = self.client.get(reverse('users:user-profile'))
+        self.assertRedirects(response, reverse('users:login') + '?next=' + reverse('users:user-profile'))
```
Comment (marjev) [inline users/tests/tests.py]: These two test methods should be merged into a single test method since they're basically testing the same thing, "what happens when an anonymous user goes to a profile page url".

Comment (marjev) [reply users/tests/tests.py]:
```
self.assertRedirects(
    response, 
    '{redirect_path}?{next_path}'.format(
        redirect_path=reverse('users:login'),
        urllib.urlencode({'next': 'reverse('users:user-profile')'})
    ))
```

does this work? you should avoid string concat when possible. If you have to do it, you should DRY the logic of concatenating into a class or utility method (like urllib does for GET arguments). Also, format method is almost always the best solution for string assembly.

File: users/tests/tests.py [hunk 26]
```diff
@@ -34,9 +44,43 @@ def test_redirection_on_register_successful(self):
...
+        response = self.client.post(reverse('users:login'), {'username': 'non_existing_user', 'password': 'pass1234'})
+        self.assertIn('Please enter a correct username and password. Note that both fields may be case-sensitive.', response.content)
+
+
+class UserProfileIntegrationTests(TestCase):
+    def test_user_profile_page_with_authenticated_user(self):
+        UserFactory(username='existing_user')
+        self.client.post(reverse('users:login'), {'username': 'existing_user', 'password': 'pass1234'})
+        response = self.client.get(reverse('users:user-profile'))
+        self.assertIn('You are logged in!', response.content)
```
Comment (marjev) [inline users/tests/tests.py]: Were you not able to use `assertTemplateUsed` instead of checking the content of the response?
This question goes to all the places you made such an assertion.

File: users/tests/tests.py [hunk 27]
```diff
@@ -34,9 +44,43 @@ def test_redirection_on_register_successful(self):
...
+class UserLoginIntegrationTests(TestCase, SimpleTestCase):
+    def test_redirection_on_login_successful(self):
+        UserFactory(username='existing_user')
+        response = self.client.post(reverse('users:login'), {'username': 'existing_user', 'password': 'pass1234'})
+        self.assertEqual(response.status_code, 302)
+
+    def test_login_successful(self):
+        UserFactory(username='existing_user')
+        response = self.client.post(reverse('users:login'), {'username': 'existing_user', 'password': 'pass1234'})
+        self.assertRedirects(response, reverse('users:user-profile'))
```
Comment (marjev) [inline users/tests/tests.py]: Merge `test_redirection_on_login_successful` and `test_login_successful` into a single test method since these two test the same scenario.

File: users/tests/tests.py [hunk 28]
```diff
@@ -34,9 +44,43 @@ def test_redirection_on_register_successful(self):
 
     def test_register_successful(self):
         response = self.client.post(reverse('users:register'), {'username': 'user_1', 'password1': 'pass1234', 'password2': 'pass1234'}, follow=True)
-        self.assertIn('You successfully created a new account!', response.content)
+        self.assertIn(UserRegistration.success_message, response.content)
```
Comment (marjev) [inline users/tests/tests.py]: If possible, use `assertTemplateUsed`.

**Review by marjev — CHANGES_REQUESTED** [review]

**Comment by mferenca** [comment]
Comments addressed, please review @marjev

File: acceptance_tests/mixins.py [hunk 29]
```diff
@@ -0,0 +1,25 @@
...
+
+
+class LoginMixin(object):
+    login_url = reverse('users:login')
+
+    def authenticate_user(self):
+        self.driver.get(urljoin(self.HOST, self.login_url))
+        self.page.username = 'TestUser'
+        self.page.password = 'Test1234'
+        self.page.login_button.click()
```
Comment (marjev) [inline acceptance_tests/mixins.py]: This method can be more reusable. If you make username and password to be parameters, you could use it in more places.

File: acceptance_tests/test_login_page.py [hunk 30]
```diff
@@ -0,0 +1,25 @@
...
+class LoginPageValidation(LoginMixin, BaseTestCase):
+    def setUp(self):
+        super(LoginPageValidation, self).setUp()
+        self.driver.get(urljoin(self.HOST, self.login_url))
+        self.page = LoginPage(self.driver)
+
+    def test_user_submitted_non_existing_username(self):
+        self.page.username = 'non_existing_user'
+        self.page.password = 'pass1234'
+        self.page.login_button.click()
```
Comment (marjev) [inline acceptance_tests/test_login_page.py]: You could use the `authenticate_user` method instead.

File: acceptance_tests/test_register_page.py [hunk 31]
```diff
@@ -1,28 +1,29 @@
...
-        self.driver.get(urljoin(self.HOST, 'users/register'))
+        super(RegisterPageValidation, self).setUp()
+        self.driver.get(urljoin(self.HOST, self.register_url))
         self.page = RegistrationPage(self.driver)
 
     def test_user_submitted_existing_username(self):
-        self.page.username = 'test'
-        self.page.password = 'pass1234'
+        self.page.username = 'TestUser'
+        self.page.password = 'Test1234'
```
Comment (marjev) [inline acceptance_tests/test_register_page.py]: How do you know that `TestUser` already exists? When was this user created?

Comment (marjev) [reply acceptance_tests/test_register_page.py]: Also, every test in this TestCase starts the same way. Create a method in this TestCase class that does that and just call it in the beginning of each test.

File: users/forms.py [hunk 32]
```diff
@@ -1,5 +1,5 @@
 from django.contrib.auth.models import User
-from django.contrib.auth.forms import UserCreationForm
+from django.contrib.auth.forms import UserCreationForm, AuthenticationForm
```
Comment (marjev) [inline users/forms.py]: nitpick: alphabetize `AuthenticationForm, UserCreationForm`

File: acceptance_tests/test_user_profile_page.py [hunk 33]
```diff
@@ -0,0 +1,25 @@
...
+    user_profile_url = reverse('users:user-profile')
+
+    def setUp(self):
+        super(UserProfilePageValidation, self).setUp()
+        self.page = LoginPage(self.driver)
+
+    def test_anonymous_user_accessing_user_profile_page(self):
+        self.driver.get(urljoin(self.HOST, self.user_profile_url))
+        page_title = self.driver.find_element_by_tag_name('h1').text
+        self.assertEqual(page_title, 'Sign in')
```
Comment (marjev) [inline acceptance_tests/test_user_profile_page.py]: Additionally, check if the login form is on the page as well.

File: acceptance_tests/test_user_profile_page.py [hunk 34]
```diff
@@ -0,0 +1,25 @@
...
+
+    def test_authenticated_user_accessing_user_profile_page(self):
+        self.authenticate_user()
+        self.driver.get(urljoin(self.HOST, self.user_profile_url))
+        message = self.driver.find_element_by_tag_name('p').text
+        self.assertEqual(message, 'You are logged in!')
```
Comment (marjev) [inline acceptance_tests/test_user_profile_page.py]: Additionally, check if the logout link is on the page as well.

**Review by marjev — CHANGES_REQUESTED** [review]

**Comment by mferenca** [comment]
comments addressed, pls review @marjev

File: acceptance_tests/mixins.py [hunk 35]
```diff
@@ -0,0 +1,23 @@
...
+from django.test import LiveServerTestCase
+from selenium import webdriver
+
+
+class BaseTestCase(LiveServerTestCase):
+    def setUp(self):
+        self.driver = webdriver.Chrome()
+
+    def tearDown(self):
+        self.driver.close()
```
Comment (marjev) [inline acceptance_tests/mixins.py]: This should live in a separate file called `testcases`. This is not a mixin.

File: acceptance_tests/mixins.py [hunk 36]
```diff
@@ -0,0 +1,23 @@
...
+
+
+class LoginMixin(object):
+    login_url = reverse('users:login')
+
+    def authenticate_user(self, username, password):
+        self.driver.get(urljoin(self.live_server_url, self.login_url))
+        self.page.username = username
+        self.page.password = password
```
Comment (marjev) [inline acceptance_tests/mixins.py]: Use `send_keys()`. http://selenium-python.readthedocs.io/navigating.html#interacting-with-the-page

Read more about OOP. https://en.wikibooks.org/wiki/A-level_Computing_2009/AQA/Problem_Solving,_Programming,_Operating_Systems,_Databases_and_Networking/Programming_Concepts/Object-oriented_programming_(OOP)

File: acceptance_tests/test_login_page.py [hunk 37]
```diff
@@ -0,0 +1,23 @@
...
+
+from page_objects import PageObject, PageElement
+
+from mixins import BaseTestCase, LoginMixin
+
+
+class LoginPage(PageObject):
+    username = PageElement(name='username')
+    password = PageElement(name='password')
+    login_button = PageElement(class_name='login-register-button')
```
Comment (marjev) [inline acceptance_tests/test_login_page.py]: You can move this into a separate file called `pages`.

File: acceptance_tests/test_login_page.py [hunk 38]
```diff
@@ -0,0 +1,23 @@
...
+
+class LoginPageValidation(LoginMixin, BaseTestCase):
+    def setUp(self):
+        super(LoginPageValidation, self).setUp()
+        self.driver.get(urljoin(self.live_server_url, self.login_url))
+        self.page = LoginPage(self.driver)
+
+    def test_user_submitted_non_existing_username(self):
+        self.authenticate_user('non_existing_user', 'pass1234')
+        error_message = self.driver.find_element_by_class_name('form-error').text
```
Comment (marjev) [inline acceptance_tests/test_login_page.py]: Login page has error message element. Page class should contain it.

Access it via page object in the test method.

File: acceptance_tests/mixins.py (hunk 36 above)
Comment (marjev) [reply acceptance_tests/mixins.py]: Do this in other tests as well.

File: acceptance_tests/test_login_page.py [hunk 39]
```diff
@@ -0,0 +1,23 @@
...
+from mixins import BaseTestCase, LoginMixin
+
+
+class LoginPage(PageObject):
+    username = PageElement(name='username')
+    password = PageElement(name='password')
+    login_button = PageElement(class_name='login-register-button')
+
+
+class LoginPageValidation(LoginMixin, BaseTestCase):
```
Comment (marjev) [inline acceptance_tests/test_login_page.py]: `LoginPageValidationTestCase` or `LoginPageTestCase`, whichever you prefer.

File: acceptance_tests/test_register_page.py [hunk 40]
```diff
@@ -1,48 +1,45 @@
...
     username = PageElement(name='username')
     password = PageElement(name='password1')
     confirmation_password = PageElement(name='password2')
-    registration_button = PageElement(class_name='submit-register')
+    registration_button = PageElement(class_name='login-register-button')
 
 
-class RegisterPageValidation(TestCase):
-    HOST = 'http://127.0.0.1:8080/'
+class RegisterPageValidation(CreateUserMixin, BaseTestCase):
```
Comment (marjev) [inline acceptance_tests/test_register_page.py]: Append `TestCase` to the name of the class.

File: acceptance_tests/test_register_page.py [hunk 41]
```diff
@@ -1,48 +1,45 @@
...
-        self.driver.get(urljoin(self.HOST, 'users/register'))
+        super(RegisterPageValidation, self).setUp()
+        self.driver.get(urljoin(self.live_server_url, self.register_url))
         self.page = RegistrationPage(self.driver)
 
-    def test_user_submitted_existing_username(self):
-        self.page.username = 'test'
-        self.page.password = 'pass1234'
-        self.page.confirmation_password = 'pass1234'
+    def fill_registration_form(self, username, password, confirmation_password):
```
Comment (marjev) [inline acceptance_tests/test_register_page.py]: Make `confirmation_password` an optional parameter and use password when it's not provided, in case you need both to exist.

File: acceptance_tests/test_user_profile_page.py [hunk 42]
```diff
@@ -0,0 +1,32 @@
+from urlparse import urljoin
+
+from django.core.urlresolvers import reverse
+
+from mixins import BaseTestCase, LoginMixin
+from test_login_page import LoginPage
+from users.tests.mixins import CreateUserMixin
+
+
+class UserProfilePageValidation(CreateUserMixin, LoginMixin, BaseTestCase):
```
Comment (marjev) [inline acceptance_tests/test_user_profile_page.py]: Append `TestCase` to the name of the class.

File: acceptance_tests/test_user_profile_page.py [hunk 43]
```diff
@@ -0,0 +1,32 @@
...
+    logout_url = reverse('users:logout')
+    user_profile_url = reverse('users:user-profile')
+
+    def setUp(self):
+        super(UserProfilePageValidation, self).setUp()
+        self.page = LoginPage(self.driver)
+
+    def test_anonymous_user_accessing_user_profile_page(self):
+        self.driver.get(urljoin(self.live_server_url, self.user_profile_url))
```
Comment (marjev) [inline acceptance_tests/test_user_profile_page.py]: DRY this and similar occurrences into a method and call it here.

File: acceptance_tests/test_login_page.py (hunk 38 above)
Comment (marjev) [reply acceptance_tests/test_login_page.py]: Propagate this change to other analogous occurrences.

File: acceptance_tests/test_user_profile_page.py [hunk 44]
```diff
@@ -0,0 +1,32 @@
...
+        page_title = self.driver.find_element_by_tag_name('h1').text
+        form = self.driver.find_element_by_tag_name('form').get_attribute('id')
```
Comment (marjev) [inline acceptance_tests/test_user_profile_page.py]: this is not a form, it's form_identifier. rename it.

File: users/tests/tests.py [hunk 45]
```diff
@@ -1,42 +1,77 @@
...
 
-class RegistrationFormUnitTests(TestCase):
-    def test_form_is_valid(self):
-        form_data = {'username': 'testUser', 'password1': 'pass1234', 'password2': 'pass1234'}
-        form = RegistrationForm(data=form_data)
+class LoginFormUnitTests(CreateUserMixin, TestCase):
+    def test_login_form_is_valid(self):
+        user = self.create_user('existing_user')
+        form = LoginForm(data={'username': user.username, 'password': self.password})
+        form.is_valid()
```
Comment (marjev) [inline users/tests/tests.py]: Why is this needed?

File: users/tests/tests.py [hunk 46]
```diff
@@ -1,42 +1,77 @@
...
+
+    def test_redirection_to_login_on_anonymous_user_access_profile_page(self):
+        response = self.client.get(reverse('users:user-profile'))
+        self.assertRedirects(
+            response,
+            '{redirect_path}?next={next_path}'.format(
+                redirect_path=reverse('users:login'),
+                next_path=reverse('users:user-profile')
+            )
+        )
```
Comment (marjev) [inline users/tests/tests.py:72]: Join the two test methods. They're testing the same thing. If you're having `assertRedirects`, you shouldn't need to test the status code.

**Review by marjev — CHANGES_REQUESTED** [review]

**Comment by mferenca** [comment]
comments addressed, pls review @marjev

**Review by marjev — APPROVED** [review]
Remember to squash the commits before merging.
